irohad/ametsuchi/impl/postgres_block_query.cpp

```diff
@@ -57,71 +60,77 @@ namespace iroha {
+    PostgresBlockQuery::getBlockIds(const std::string &account_id) {
       std::vector<uint64_t> block_ids;
-      client_.smembers(account_id, [&block_ids](cpp_redis::reply &reply) {
-        for (const auto &block_reply : reply.as_array()) {
-          const auto &string_reply = block_reply.as_string();
-
-          // check if reply is an integer
-          if (isdigit(string_reply.c_str()[0])) {
-            block_ids.push_back(std::stoul(string_reply));
-          }
-        }
-      });
-      client_.sync_commit();
-      return block_ids;
+      return execute_(
+                 "SELECT DISTINCT height FROM height_by_account_set WHERE "
+                 "account_id = '"
+                 + account_id + "';")
+                 | [&block_ids](const auto &result)
+                 -> std::vector<iroha::model::Block::BlockHeightType> {
+        return transform<iroha::model::Block::BlockHeightType>(
+            result, [&](const auto &row) {
+              pqxx::binarystring height_reply(row.at("height"));
+              auto height = std::stoul(height_reply.str());
+              return height;
+            });
+      };
     }
 
     boost::optional<iroha::model::Block::BlockHeightType>
-    RedisBlockQuery::getBlockId(const std::string &hash) {
+    PostgresBlockQuery::getBlockId(const std::string &hash) {
       boost::optional<uint64_t> blockId;
-      client_.get(hash, [&blockId](cpp_redis::reply &reply) {
-        if (reply.is_null()) {
-          blockId = boost::none;
-        } else {
-          blockId = std::stoul(reply.as_string());
-        }
-      });
-      client_.sync_commit();
-
-      return blockId;
+      return execute_("SELECT height FROM height_by_hash WHERE hash = "
+                      + transaction_.quote(
+                            pqxx::binarystring(hash.data(), hash.size()))
+                      + ";")
+                 | [&](const auto &result)
+                 -> boost::optional<iroha::model::Block::BlockHeightType> {
+        return result.size() == 0 ? boost::none
+                                  : boost::optional<uint64_t>(std::stoul(
+                                        result[0].at("height").c_str()));
+      };
     }
 
-    std::function<void(cpp_redis::reply &)> RedisBlockQuery::callbackToLrange(
-        const rxcpp::subscriber<model::Transaction> &s, uint64_t block_id) {
-      return [this, &s, block_id](cpp_redis::reply &reply) {
-        auto tx_ids_reply = reply.as_array();
+    std::function<void(pqxx::result &result)> PostgresBlockQuery::callback(
+        const rxcpp::subscriber<model::Transaction> &subscriber,
+        uint64_t block_id) {
+      return [this, &subscriber, block_id](pqxx::result &result) {
+        std::vector<std::string> tx_ids_reply;
+        tx_ids_reply.reserve(result.size());
+        for (pqxx::result::const_iterator i = result.begin(); i != result.end();
+             ++i)
+          tx_ids_reply.push_back((*i).at("index").as<std::string>());
```

Comment:
This part seems too complicated, because these ids are later converted to `unsigned long`. So this conversion part can be removed, and the last lambda (`[&](const auto &block) {`) can be changed.
The following seems easier to read:
```C++
boost::for_each(result | boost::adaptors::transformed([](const auto &x){
  return x.at("index").as<size_t>();
}), [&](auto x){
  const auto &tx = block.transactions.at(x);
  subscriber.on_next(tx);
});
```